### PR TITLE
i#5843 scheduler: Implement speculation with nops

### PR DESCRIPTION
**Pull request by derekbruening**
Implements initial speculation support, supplying nops.  Speculation is separated into its own class where we can fill in different strategies in the future.

The start_speculation() function takes a flag controlling whether the scheduler queues up the current record and re-returns it as the first record after speculation stops.  This is often what a simulator wants as it has to read the instruction record following a branch to determine whether it is on the wrong path, and it would like to resume with that already-read instruction after speculation.

Adds a unit test.

Issue: #5843

### CODE REVIEW

File: clients/drcachesim/scheduler/scheduler.h [hunk 1]
```diff
@@ -480,16 +481,28 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
...
-         * Stops speculative execution and resumes the regular stream of records
-         * from the point at which the prior start_speculation call was made.
-         * Returns STATUS_INVALID if there was no prior start_speculation() call.
+         * Stops speculative execution and resumes the regular stream of records from
+         * the point at which the most distant prior start_speculation() call without an
+         * intervening stop_speculation() call was made (either repeating the current
+         * record at that time, if "true" was passed for "queue_current_record" to
+         * start_speculation(), or continuing on the subsequent record if "false" was
+         * passed).  Returns #dynamorio::drmemtrace::scheduler_tmpl_t::STATUS_INVALID if
+         * there was no prior start_speculation() call.
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.h:505]: or if stop_speculation was already called.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 2]
```diff
@@ -796,6 +796,126 @@ test_synthetic()
...
+        case 4:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 2);
+            // We realize now that we mispredicted that the branch would be taken.
+            // We ask to queue this record for post-speculation.
+            stream->start_speculation(100, true);
+            break;
+        case 5:
+            // We should now see nops from the speculator.
+            assert(type_is_instr(memref.instr.type));
```
Comment (abhinav92003) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:868]: This doesn't verify that it's a nop. Maybe we can have a hard-coded check here for the encoding.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 3]
```diff
@@ -796,6 +796,126 @@ test_synthetic()
...
+        case 8:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 3);
+            break;
+        case 9:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 4);
+            // We realize now that we mispredicted that the branch would be taken.
+            // This time we do *not* ask to queue this record for post-speculation.
+            stream->start_speculation(200, false);
```
Comment (abhinav92003) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:897]: Maybe also test multiple start_speculation calls to confirm that PC is changed after each?

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:868]: I tried linking in the decoder to decode and run instr_is_nop() and that worked w/o too much effort so I'll go with that.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:897]: Done.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:883]: Maybe pulling in the decoder was a mistake: dealing with Windows toolchain issues due to that...

**Pull request merged**